pkg/util/http.go

```diff
@@ -374,3 +392,49 @@ func copyValues(src url.Values) url.Values {
 func IsHTTPStatusCode(code codes.Code) bool {
 	return int(code) >= 100 && int(code) < 600
 }
+
+// BufferProvider provides request body and decoding buffers.
+type BufferProvider struct {
+	size     int
+	body     *bytes.Buffer
+	decoding *bytes.Buffer
+}
+
+// NewBufferProvider returns a new BufferProvider given an initial buffer size.
+func NewBufferProvider(size int) BufferProvider {
+	return BufferProvider{
+		size: size,
+	}
+}
+
+// BodyBuffer gets the buffer for reading the request body.
+func (p *BufferProvider) BodyBuffer() *bytes.Buffer {
```

Comment:
I would suggest to handle the case when `p == nil` in this method (by always returning new buffer).

I would also suggest to pass expected size directly here to this method, to have "grow" in single place only. (If expected size is unknown, pass 0)